services/hanke-service/src/test/kotlin/fi/hel/haitaton/hanke/email/EmailSenderServiceTest.kt

```diff
@@ -0,0 +1,571 @@
+    private val mailSender: JavaMailSender = spyk()
+    private val emailSenderService = EmailSenderService(mailSender, emailConfig, featureFlags)
+
+    @BeforeEach
+    fun clean() {
+        clearAllMocks()
+        every { mailSender.createMimeMessage() } returns MimeMessage(null as Session?)
+    }
+
+    @Nested
+    inner class HankeInvitation {
+        private val invitationToken = "MgtzRbcPsvoKQamnaSxCnmW7"
+        private val hankeInvitation =
+            HankeInvitationData(
+                inviterName = INVITER_NAME,
+                inviterEmail = INVITER_EMAIL,
+                recipientEmail = TEST_EMAIL,
+                hankeTunnus = HANKE_TUNNUS,
+                hankeNimi = HANKE_NIMI,
+                invitationToken = invitationToken,
+            )
+
+        private fun sendAndCapture(): MimeMessage {
+            val email = slot<MimeMessage>()
+            justRun { mailSender.send(capture(email)) }
+
+            emailSenderService.sendHankeInvitationEmail(hankeInvitation)
+
+            return email.captured
+        }
+
+        @Test
+        fun `has the correct subject`() {
+            val email = sendAndCapture()
+
+            assertThat(email.subject)
+                .isEqualTo(
+                    "Haitaton: Sinut on lisätty hankkeelle HAI24-1 " +
+                        "/ Du har lagts till i projektet HAI24-1 " +
+                        "/ You have been added to project HAI24-1"
+                )
+        }
+
+        @Test
+        fun `has a header line in the body`() {
+            val (textBody, htmlBody) = sendAndCapture().bodies()
+
+            assertThat(textBody)
+                .contains(
+                    "Sinut on lisätty hankkeelle HAI24-1 " +
+                        "/ Du har lagts till i projektet HAI24-1 " +
+                        "/ You have been added to project HAI24-1"
+                )
+            assertThat(htmlBody)
+                .containsEscaped(
+                    "Sinut on lisätty hankkeelle HAI24-1 " +
+                        "/ Du har lagts till i projektet HAI24-1 " +
+                        "/ You have been added to project HAI24-1"
+                )
+        }
+
+        @Nested
+        inner class BodyInFinnish {
+            @Test
+            fun `contains the invitation url`() {
+                val (textBody, htmlBody) = sendAndCapture().bodies()
+
+                assertThat(textBody)
+                    .contains("https://haitaton.hel.fi/fi/kutsu?id=$invitationToken")
+                assertThat(htmlBody)
+                    .contains("""<a href="https://haitaton.hel.fi/fi/kutsu?id=$invitationToken">""")
+            }
+
+            @Test
+            fun `contains the signature`() {
+                val (textBody, htmlBody) = sendAndCapture().bodies()
+
+                assertThat(textBody).all {
+                    containsLine("Tämä on automaattinen sähköposti – älä vastaa tähän viestiin.")
```

Comment:
Maybe something could be done about the code quality report (duplication)? I like the explicit style here, but it does make the sonar report look worse than it actually is. I will leave this up to your consideration.

Reply:
>Programs must be written for people to read, and only incidentally for machines to execute.
― Harold Abelson, Structure and Interpretation of Computer Programs

I do believe this quote is on to something. And I think static code analysis falls under the "machine" category.

Still, I refactored the tests to have less duplication in them. See how you like them.

Reply:
> Programs must be written for people to read (and maintain).
-- Me

This made sonar happy also. 👍